File: server/chainup_agent/api/schemas/admin_orchestration.py
"""Admin orchestration API schemas — 运行场景 · 执行策略."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

FailureStopPolicy = Literal["stop_notify", "stop_silent", "escalate_manual"]


class OrchestrationExecutionPolicyOut(BaseModel):
    model_config = ConfigDict(populate_by_name=True)

    orchestration_registry_version: str = Field(alias="orchestrationRegistryVersion")
    auto_execution_allowed: bool = Field(alias="autoExecutionAllowed")
    block_auto_high_risk_write: bool = Field(alias="blockAutoHighRiskWrite")
    query_auto_default: bool = Field(alias="queryAutoDefault")
    confirmation_required_for_writes: bool = Field(alias="confirmationRequiredForWrites")
    second_confirm_large_notional: bool = Field(alias="secondConfirmLargeNotional")
    high_leverage_confirm: bool = Field(alias="highLeverageConfirm")
    max_notional_usdt: int = Field(alias="maxNotionalUsdt")
    max_leverage: int = Field(alias="maxLeverage")
    max_daily_write_operations: int = Field(alias="maxDailyWriteOperations")
    rate_limit_note: str = Field(alias="rateLimitNote")
    max_retries: int = Field(alias="maxRetries")
    execution_timeout_seconds: int = Field(alias="executionTimeoutSeconds")
    failure_stop_policy: FailureStopPolicy = Field(alias="failureStopPolicy")
    max_tool_calls: int = Field(alias="maxToolCalls")
    max_orchestration_steps: int = Field(alias="maxOrchestrationSteps")
    max_model_rounds: int = Field(alias="maxModelRounds")
    model_rounds_limit_enabled: bool = Field(alias="modelRoundsLimitEnabled")
    budget_exceeded_stable_code: str = Field(alias="budgetExceededStableCode")
    c_class_pool_note: str = Field(alias="cClassPoolNote")
    retry_summary: str = Field(alias="retrySummary")
    unknown_handling_summary: str = Field(alias="unknownHandlingSummary")
    engineering_spec_refs: list[str] = Field(default_factory=list, alias="engineeringSpecRefs")

File: admin/src/shared/api/admin-orchestration.ts
// 作者: 杨永的Agent
// 日期: 2026-05-20
// 修改功能: **GET /api/v1/admin/orchestration/policy** 执行策略读模型（FE_HANDOFF · Phase2 运行编排）

import { HTTPError } from 'ky'

import { AppError } from '@/shared/api/errors'
import { httpClient, toAppError } from '@/shared/api/http-client'

const PREFIX = 'v1/admin/orchestration'

export type OrchestrationFailureStopPolicy = 'stop_notify' | 'stop_silent' | 'escalate_manual'

/** 对齐后端 `OrchestrationExecutionPolicyOut` */
export interface OrchestrationExecutionPolicy {
  orchestrationRegistryVersion: string
  autoExecutionAllowed: boolean
  blockAutoHighRiskWrite: boolean
  queryAutoDefault: boolean
  confirmationRequiredForWrites: boolean
  secondConfirmLargeNotional: boolean
  highLeverageConfirm: boolean
  maxNotionalUsdt: number
  maxLeverage: number
  maxDailyWriteOperations: number
  rateLimitNote: string
  maxRetries: number
  executionTimeoutSeconds: number
  failureStopPolicy: OrchestrationFailureStopPolicy
  maxToolCalls: number
  maxOrchestrationSteps: number
  maxModelRounds: number
  modelRoundsLimitEnabled: boolean
  budgetExceededStableCode: string
  cClassPoolNote: string
  retrySummary: string
  unknownHandlingSummary: string
  engineeringSpecRefs: string[]
}

async function normalizeKyError(error: unknown): Promise<never> {
  if (error instanceof HTTPError) {
    const cloned = error.response.clone()
    const body: unknown = await cloned.json().catch(() => null)
    if (
      body !== null &&
      typeof body === 'object' &&
      'message' in body &&
      typeof (body as { message: unknown }).message === 'string'
    ) {
      const code =
        'code' in body && typeof (body as { code: unknown }).code === 'string'
          ? (body as { code: string }).code
          : undefined
      throw new AppError((body as { message: string }).message, {
        status: error.response.status,
        code,
      })
    }
  }
  throw await toAppError(error)
}

export async function getOrchestrationPolicy(): Promise<OrchestrationExecutionPolicy> {
  try {
    return await httpClient.get(`${PREFIX}/policy`).json<OrchestrationExecutionPolicy>()
  } catch (e) {
    throw await normalizeKyError(e)
  }
}
